File: algorithms/dynamic_programming/72_EditDistance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int l1 = word1.length();
        int l2 = word2.length();
        int[][] dp = new int[l1+1][l2+1];
        for (int i=1; i<l1+1;++i){
            dp[i][0] = i;
        }
        for (int i=1; i<l2+1;++i){
            dp[0][i] = i;
        }
        
        for (int i=1; i<l1+1; ++i) {
            for (int j = 1; j<l2+1; ++j) {
                /**
                 * If current char of word1 and word2 match, then dp[i][j] simply equals to dp[i-1][j-1], if they don't match
                 * suppose current char of word1 is c, current char of word2 is d, there are three options how we can change
                 * word1[:i] to word2[:j]
                 * 
                 * 1. We replace c with d, so dp[i][j] = dp[i-1][j-1]+1
                 * 2. We add d to the end, so dp[i][j] = dp[i][j-1]+1
                 * 3. We delete c from the end, so dp[i][j] = dp[i-1][j]+1
                 * 
                 * Then we can simply use the minimum of these 3 as the dp[i][j] value
                 */
                int cross = dp[i-1][j-1];
                if (word1.charAt(i-1) != word2.charAt(j-1))
                    ++cross;
                
                int delete = dp[i-1][j] + 1;
                int add = dp[i][j-1] + 1;
                int min = Math.min(add, delete);
                min = Math.min(min, cross);
                dp[i][j] = min;
            }
        }
        return dp[l1][l2];
    }
}

File: algorithms/dynamic_programming/72_EditDistance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)

        # just to make sure we use minimum possible space
        if m < n:
            word1, word2, m.n = word2, word1, n, m

        # O(mn) space solution
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m + 1):
            dp[i][0] = i

        for j in range(n + 1):
            dp[0][j] = j

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                cross = dp[i - 1][j - 1]
                if word1[i - 1] != word2[j - 1]:
                    cross += 1

                add = dp[i - 1][j] + 1
                delete = dp[i][j - 1] + 1
                dp[i][j] = min(add, delete, cross)

        # O(2 * min(m, n)) space solution
        dp2 = [0] * (n + 1)
        for j in range(n + 1):
            dp2[j] = j
        for i in range(1, m + 1):
            next = [i] + [0] * (n)
            for j in range(1, n + 1):
                cross = dp2[j - 1]
                if word1[i - 1] != word2[j - 1]:
                    cross += 1
                add = dp2[j] + 1
                delete = next[j - 1] + 1
                next[j] = min(cross, add, delete)
            dp2 = next

        # O(min(m, n)) space solution
        dp3 = [0] * (n + 1)
        for j in range(n + 1):
            dp3[j] = j
        for i in range(1, m + 1):
            prev = dp3[0]
            dp3[0] = i
            for j in range(1, n + 1):
                temp = dp3[j]
                cross = prev
                if word1[i - 1] != word2[j - 1]:
                    cross += 1

                add = dp3[j] + 1
                delete = dp3[j - 1] + 1
                dp3[j] = min(cross, add, delete)
                prev = temp

        return dp3[-1]